docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md

```diff
@@ -0,0 +1,245 @@
+# Build-time and runtime detection for hardware acceleration with advance CPU features
+
+This document describes how to support the algorithms that can do hardware acceleration with advance CPU features.
+
```

Comment:
I do not know that.  I will take a look. 
Where is document about Mbed TLS 4.0?

Reply:
>If a driver has fallback disabled, the subsequent drivers in the list are ignored for this algorithm.

Do the subsequent drivers exist in the target binary? 

How to reduce code size with one driver per algorithm in PSA drivers? This case is for known SOC.

I did not catch above points in [PSA drivers](https://github.com/Mbed-TLS/mbedtls/blob/development/docs/proposed/psa-driver-interface.md)

Reply:
You select which implementations to include by passing a set of JSON files to `generate_driver_wrappers.py`. If you don't want to include a driver, don't pass its JSON file. Furthermore, if there are multiple drivers for an algorithm, only the first one is included (or more precisely: the build only includes the drivers up to the first one that has fallback=false).

Reply:
> Do we want the legacy mechanism to last for a long time? Or are we going to remove it in Mbed TLS 4.0?

We want to retire the legacy mechanism as soon as possible, however we only should do it when
1. the PSA Crypto Driver interface can do the same job
2. all use cases in TLS and X509 can work in PSA Crypto mode

I am not completely up to date on the progress in these areas, but I think we are pretty close to achieving 2. (if not already there), but there is some more work to be done to achieve 1. If we can get there by 4.0, I think we should retire the legacy mechanism in that release.

> How will the legacy mechanism work for PSA-only users?

Since they won't coexist for very long, I think we should go for the option that is most convenient for us and presents an at least marginally sensible interaction between the two mechanisms.

> The single feature detection function in this proposal makes more sense for core CPU features.

I am not sure about that, I always thought that these crypto specific extensions that are targeted at very specific algorithms/modules (like AES and SHA for example) would eventually be moved into a standard PSA Crypto Driver that happens to be supplied with the library. (Although in this case we still can have a single feature detection function, which would be called from the drivers.)

Reply:
Should I abandon this proposal ? That really needs more efforts to implement it

But I will add `MBEDTLS_RUNTIME_C` and I think it can be reused in PSA drivers.

Reply:
> Should I abandon this proposal ? That really needs more efforts to implement it

I don't think we should abandon this proposal. What would need more effort to implement?

Reply:
We should change sha256/512 and aes for x86 also. 

I just take a look at those module again. 
- AES for x86/x64, it implemented with assembly, that is simple, just replace runtime detection function.
- SHA512/256,  runtime detection function can be replace firstly. I have no idea about effort on other part.

Reply:
I have update #7078 to remove many configuration options.